### PR TITLE
Editing Plan Frontend: Adding new semesters

### PR DESCRIPTION
**Pull request by DaleCS**
Modified the function that handles adding new semesters from the old edit plan branch and then moved it into the 023-elanaedit. User can now add new semesters.

## Summary of Changes:
1. isAddingSemester property for each year has been changed into an object called newSemesterWidget. This object holds the boolean flag value used to toggle the new semester widget and a string value to hold the value of the term select menu.
`
newSemesterWidget: {
     active: boolean,
     termSelect: string
}
`
prototype8

2. Added the addNewSemester() in plan.service. It handles manipulating the plan variable in s-degree-plan-editor component to add a new semester.

3. Also included sortSemesterChronologically function. It sorts semester chronologically. It should be called everytime a semester is added/deleted.

### CODE REVIEW

File: client/src/app/plan.service.ts [hunk 1]
```diff
@@ -99,6 +99,57 @@ export class PlanService {
     return this.user.pipe(map(mapCallback));
   }
 
+  /**
+   * Adds a new semester into years chronologically. If term already exists, nothing is done.
+   * @param term The term in string form
+   * @param yearIndex  the Index of the year that the semester will be added to
+   * @param years The years array
+   */
+  addNewSemester(term: string, yearIndex: number, years: Array<any>): void {
```
Comment (DaleCS) [inline client/src/app/plan.service.ts:107]: This function handles changing the plan variable in s-degree-plan-editor to add a new semester.

File: client/src/app/plan.service.ts [hunk 2]
```diff
@@ -99,6 +99,57 @@ export class PlanService {
...
+      window.alert("Semester already exists!");
+    }
+  }
+
+  /**
+   * Chronologically sorts the semesters at the given index of years
+   * @param yearIndex The index of the year in the year array
+   * @param years The years array
+   */
+  private sortSemestersChronologically(
```
Comment (DaleCS) [inline client/src/app/plan.service.ts:140]: Helper function that will be reused by other functions. It sorts the semesters chronologically. (Summer, Fall, Winter, Spring)

File: client/src/app/s-degree-plan-editor/s-degree-plan-editor.component.html [hunk 3]
```diff
@@ -176,33 +178,37 @@
...
+              class="mr8"
+              *ngIf="year.newSemesterWidget.active === true"
+            >
               <mat-label>Semester</mat-label>
-              <mat-select>
-                <mat-option value="SUMMER">Summer</mat-option>
-                <mat-option value="FALL">Fall</mat-option>
-                <mat-option value="SPRING">Spring</mat-option>
-                <mat-option value="WINTER">Winter</mat-option>
+              <mat-select [(ngModel)]="year.newSemesterWidget.termSelect">
```
Comment (DaleCS) [inline client/src/app/s-degree-plan-editor/s-degree-plan-editor.component.html:207]: Used ngModel to connect the select menu to the newSemesterWidget property in the plan variable of s-degree-plan-editor component.

File: client/src/app/s-degree-plan-editor/s-degree-plan-editor.component.ts [hunk 4]
```diff
@@ -30,9 +30,12 @@ export class SDegreePlanEditorComponent implements OnInit {
       this.plan.user = "";
       this.plan.program = "";
       this.plan.isAddingYear = false;
-      this.plan.years.forEach(
-        (semester) => (semester.isAddingSemester = false)
-      );
+      this.plan.years.forEach((year) => {
```
Comment (DaleCS) [inline client/src/app/s-degree-plan-editor/s-degree-plan-editor.component.ts:33]: I realized that I need to change isAddingSemester into an object that contains both a boolean flag (for toggling the add semester widget) and then a string property to hold the value of the select menu of terms.

File: client/src/app/s-degree-plan-editor/s-degree-plan-editor.component.ts [hunk 5]
```diff
@@ -257,9 +260,21 @@ export class SDegreePlanEditorComponent implements OnInit {
 
   ngOnInit() {}
 
+  /**
+   * Handles click event on "Add Semester" button. Adds a new semester to plan
+   * @param yearIndex The
+   */
+  onClickAddNewSemester(yearIndex: number): void {
```
Comment (DaleCS) [inline client/src/app/s-degree-plan-editor/s-degree-plan-editor.component.ts:267]: onClick handler function in the component that calls planService.

File: client/src/app/plan.service.ts [hunk 6]
```diff
@@ -98,6 +98,57 @@ export class PlanService {
...
```
Comment (ellamaolson) [inline client/src/app/plan.service.ts:140]: Add return type to function

Comment (ellamaolson) [reply client/src/app/plan.service.ts:140]: Also for this function  private calculateSemesterStatistics(semester: Semester) {

File: client/src/app/s-degree-plan-editor/s-degree-plan-editor.component.html [hunk 7]
```diff
@@ -27,7 +27,9 @@
               />
               <div class="semester-card">
                 <div class="semester-card-term-and-status">
-                  <span> {{ semester.term }} {{ semester.year }}</span>
+                  <span>
```
Comment (ellamaolson) [inline client/src/app/s-degree-plan-editor/s-degree-plan-editor.component.html]: check if semester term and year are defined with *ngIf in this span

Comment (ellamaolson) [reply client/src/app/s-degree-plan-editor/s-degree-plan-editor.component.html]: likewise, add to line 13 
<span class="year-range">{{ year.beginning }} - {{ year.ending }}</span>

and line 180 in the existing ngIf
<button
              *ngIf="year.newSemesterWidget.active === false"
              mat-button
              (click)="toggleAddingSemester(yearIndex)"
              class="add-semester-font-clr w100"
            >
              + ADD A SEMESTER FOR {{ year.beginning }} - {{ year.ending }}
            </button>

**Review by ellamaolson — CHANGES_REQUESTED** [review]

**Comment by ellamaolson** [comment]
Very nice summary and preview, it was really helpful in just reviewing the entire PR on github!

File: client/src/app/s-degree-plan-editor/s-degree-plan-editor.component.html (hunk 7 above)
Comment (DaleCS) [reply client/src/app/s-degree-plan-editor/s-degree-plan-editor.component.html]: Added

File: client/src/app/plan.service.ts (hunk 6 above)
Comment (DaleCS) [reply client/src/app/plan.service.ts:140]: Added

**Review by ellamaolson — APPROVED** [review]

**Pull request merged**